consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::{
+    fairness::{
+        conflict_key::ConflictKeyRegistry, conflict_zone::ConflictZone, pending_zone::PendingZone,
+    },
+    TransactionShuffler,
+};
+use aptos_types::transaction::SignedTransaction;
+use itertools::zip_eq;
+use selection_tracker::SelectionTracker;
+use std::collections::BTreeSet;
+
+pub(crate) mod conflict_key;
+mod conflict_zone;
+mod pending_zone;
+mod selection_tracker;
+
+#[cfg(test)]
+mod tests;
+
+type TxnIdx = usize;
+
+const NUM_CONFLICT_ZONES: usize = 3;
+
+#[derive(Debug)]
+struct FairnessShuffler {
+    sender_conflict_window_size: usize,
+    module_conflict_window_size: usize,
+    entry_fun_conflict_window_size: usize,
+}
+
+impl TransactionShuffler for FairnessShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let conflict_key_registries = ConflictKeyRegistry::build_registries(&txns);
+        let order = FairnessShufflerImpl::new(self, &conflict_key_registries).shuffle();
+        reorder(txns, &order)
+    }
+}
+
+fn reorder<T: Clone>(txns: Vec<T>, order: &[TxnIdx]) -> Vec<T> {
+    assert_eq!(txns.len(), order.len());
+    order.iter().map(|idx| txns[*idx].clone()).collect()
+}
+
+struct FairnessShufflerImpl<'a> {
+    conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    conflict_zones: [ConflictZone<'a>; NUM_CONFLICT_ZONES],
+    pending_zones: [PendingZone<'a>; NUM_CONFLICT_ZONES],
+    selected_order: Vec<TxnIdx>,
+    selection_tracker: SelectionTracker,
+}
+
+impl<'a> FairnessShufflerImpl<'a> {
+    pub fn new(
+        shuffler: &FairnessShuffler,
+        conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    ) -> Self {
+        let num_txns = conflict_key_registries[0].num_txns();
+        assert!(conflict_key_registries
+            .iter()
+            .skip(1)
+            .all(|r| r.num_txns() == num_txns));
+
+        Self {
+            conflict_key_registries,
+            selected_order: Vec::with_capacity(num_txns),
+            selection_tracker: SelectionTracker::new(num_txns),
+            conflict_zones: ConflictZone::build_zones(conflict_key_registries, [
+                shuffler.sender_conflict_window_size,
+                shuffler.module_conflict_window_size,
+                shuffler.entry_fun_conflict_window_size,
+            ]),
+            pending_zones: PendingZone::build_zones(conflict_key_registries),
+        }
+    }
+
+    pub fn shuffle(mut self) -> Vec<TxnIdx> {
```

Comment:
Can you clarify what is the average and worst case time complexity for this?

Reply:
coming up.